enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,742 @@
+					skip_update_of_resolvedName
+				else
+					combine_ips(resolvedName.ips, ips)
+			else
+				resolvedName.ips = ips
+			resolvedName.ttl = ttl
+			resolvedName.lastLookupTime = now()
+			resolvedName.resolutionFailures = 0
+			cond = condition{type: "Degraded", status: false, reason: success_reason(response), message: success_message(response)}
+			add_condition(resolvedName.conditions, cond)
+
+		else if is_wildcard(obj.spec.name) && resolvedName.dnsName == obj.spec.name && is_regular(dnsName)
+			&& match_next_lookup_time(resolvedName.lastLookupTime, resolvedName.ttl, ttl) && match_ips(resolvedName.ips, ips):
+			skip_update_of_resolvedName
+		
+		else if is_wildcard(dnsName)
+			&& match_next_lookup_time(resolvedName.lastLookupTime, resolvedName.ttl, ttl) && match_ips(resolvedName.ips, ips):
+			remove(obj.status.resolvedNames, resolvedName)
+	
+	update_status_dns_name_resolver(obj)
+```
+
+* If a regular DNS name in the lookup matches with a `DNSNameResolver` CR corresponding to a regular DNS name, then the `.status` of
+the `DNSNameResolver` CR will be updated with the DNS name and the corresponding current IP addresses along with the TTL and the current time
+as the last lookup time.
+* If a regular DNS name in the lookup matches with a `DNSNameResolver` CR corresponding to a wildcard DNS name, then the `.status` of
+the `DNSNameResolver` CR will be updated, if the IP addresses received in the response doesn't match with the IP addresses and the next lookup time
+corresponding to the wildcard DNS name. Otherwise, the plugin will update the `.status` with the regular DNS name and the corresponding current IP addresses
+along with the TTL and the next lookup time based on the TTL.
+* If the DNS lookup is for a wildcard DNS name and it matches with the `DNSNameResolver` CR corresponding to the wildcard DNS name,
+then the `ocp_dnsnameresolver` plugin will update the `.status` of the corresponding `DNSNameResolver` CR with the wildcard DNS name
+and the corresponding current IP addresses along with the TTL and the current lookup time as the last lookup time. If the wildcard DNS name's
+corresponding IP addresses and next lookup time (TTL + last lookup time) matches that of any other regular DNS name, added to the `.status`
+of the `DNSNameResolver` CR, then the details of the regular DNS name will be removed.
+* However, the updates will take place if there is a change in the next lookup time (TTL + last lookup time) for the DNS name or if the next
+lookup time is same for the DNS name, but the current IP addresses are different from the existing IP addresses. For the latter, the current
+IP addresses are added to the existing IP addresses of the DNS name. This will take care of the scenario where different CoreDNS pods gets
+different subsets of the IP addresses in the response to the DNS lookup of the same DNS name. The next time to lookup will be same for them
+though the IP addresses may differ due to upstream DNS load balancing.
```

Comment:
let's say I have current status with 
```
ips: [1.1.1.1]
TTLSecond: 10
```
and a new DNS reply for the same DNSName with `ip=1.1.1.2, TTL=600`, if I got that right, the status will be updated to
```
ips: [1.1.1.1, 1.1.1.2]
TTLSecond: 600
```
but that TTL is not correct for the first ip, and let's consider the first ip is actually not used anymore, so it needs to be deleted in 10 seconds. How will this work?

Reply:
The update will take place in 2 scenarios:

- If the next lookup times are different (TTL + last lookup time)
  For example, if the status is:
  ```
  ips: [1.1.1.1]
  ttlSecond: 10
  lastLookupTime: 12:00
  ```
  And a new lookup happens at 12:10 with `ips=1.1.1.2, ttl=60`, then the next lookup time is different for the IP addresses. So the status will be updated to:
  ```
  ips: [1.1.1.2]
  ttlSecond: 60
  lastLookupTime: 12:10
  ```
- If the next lookup time are same but IPs are different
  If we again have the same status:
  ```
  ips: [1.1.1.1]
  ttlSecond: 10
  lastLookupTime: 12:00
  ```
  And a new lookup happens at 12:05 with `ips=1.1.1.2, ttl=5`, then the next lookup time for both the IP addresses are same (12.10). So the status will be updated to:
  ```
  ips: [1.1.1.1, 1.1.1.2]
  ttlSecond: 5
  lastLookupTime: 12:05
  ```

Reply:
not sure I got the first part, so if a new lookup happens at 12:05 instead of 12:10, what would be the result?
Another example I would like to understand is what happens if the status is
```
ips: [1.1.1.1]
ttlSecond: 60
lastLookupTime: 12:00
```
and a new lookup happens at 12:10 with TTL=10 and a new ip 1.1.1.2?
I guess some general algorithm for this merging would be nice to have.

Reply:
It will match the 1st scenario, where the next time to lookup is different. The status in both the cases that you have mentioned will be updated with the latest IP and TTL. So for the example that I had given:
```
ips: [1.1.1.2]
ttlSecond: 60
lastLookupTime: 12:05
```
For the example that you had given:
```
ips: [1.1.1.2]
ttlSecond: 10
lastLookupTime: 12:10
```
In the scenario that different CoreDNS pods get different sets of IPs, the TTL of those IP addresses will be the same. That's what is being captured in the second scenario mentioned in my previous comment. Thus if the next time to lookup is same for the IPs obtained by different CoreDNS pods, then the IPs will be combined.

However, if the next time to lookup is different, it means that new IPs have been updated in the DNS servers and the existing IPs and TTL have to updated in the status. This is what happens in the 2 examples that you asked. The  next time to lookup does not match and thus the status is updated with the new IP and TTL.

I have already added the details in the pseudocode here (https://github.com/openshift/enhancements/pull/1335/files/6854ffc27aaa56a0b48ba0d01a8a5cb59a151ffe#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R361-R373) and explained it here (https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R409-R413).

Reply:
thanks for the links, I knew there was some explanation. I have some concerns though:
1. the concept of the "same time": we only have time in seconds, which means rounding errors will happen
2. "In the scenario that different CoreDNS pods get different sets of IPs, the TTL of those IP addresses will be the same" how do you know that? Is it guaranteed by some standard? I think I've seen examples of when this is not the case, and dns responses may have both different ips and next lookup time (now+TTL)

Reply:
> 1. the concept of the "same time": we only have time in seconds, which means rounding errors will happen

The time matching issue is explained here: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R414-R416



> 2\. "In the scenario that different CoreDNS pods get different sets of IPs, the TTL of those IP addresses will be the same" how do you know that? Is it guaranteed by some standard? I think I've seen examples of when this is not the case, and dns responses may have both different ips and next lookup time (now+TTL)

The RFC (https://datatracker.ietf.org/doc/html/rfc2181#section-5.2) clarifies that the TTL of a record will be same for all IPs. Thus, whenever a partial set of IPs are received by different CoreDNS pods, the TTL values will be same.

Reply:
So I have tested this with one of the dns names we had problems with before (`sts.amazonaws.com`) and I could get 2 different replies with 2 different TTLs if I resolve this name from the local machine and the cluster node. You can probably reproduce it the same way, I am not sure if that means that DNS server is not working correctly or if these record are not considered the same RRSet

Reply:
Different nameservers may respond with different IPs (and TTL) for the same DNS name. The nameservers will be configured in the `/etc/resolv.conf` file. As the CoreDNS pods in a cluster are configured with the same nameserver, there won't be any disparity in the information. You can check how resolution of the external DNS names is configured in CoreDNS: https://github.com/openshift/cluster-dns-operator/blob/master/README.md#how-it-works

Reply:
I think I was told before that there is no guarantee that all nodes in the cluster are configured with the same `resolv.conf` (like it is the config by default, but it can be changed, and that was a part of the problem).
Also, I am not sure the doc you linked says that all `resolv.conf` must be the same on all cluster nodes?

Reply:
> think I was told before that there is no guarantee that all nodes in the cluster are configured with the same `resolv.conf` (like it is the config by default, but it can be changed, and that was a part of the problem).

We can add in the documentation that if all the nodes are not configured with the same nameserver in the `/etc/resolv.conf` then there are chances that the EgressFirewall won't work correctly. It is already documented in current EgressFirewall documentation that pods should use the same local nameserver as that of the egress firewall controller: https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#domain-name-server-resolution_configuring-egress-firewall-ovn

Reply:
I had a discussion with @Miciah last week. Here is a summary of the discussion:
- Even though the nameservers are same on all nodes, if the cluster is geo-distributed, the nameservers may use different IPs in different regions. The assumption that the TTLs will be same may not hold.
- Removing IPs immediately after the expiration of the TTL may cause existing connections to drop. We can wait for a grace period (say 1s) after the TTL expiration to remove an IP.
- When the `DNSNameResolver` sends the request for DNS lookup it can send it to a maximum of 5 random CoreDNS pods rather than the CoreDNS service. This may ensure that the IPs for a DNS name are from different nameservers/regions.

I will update the EP based on this discussion.

Reply:
> The RFC (https://datatracker.ietf.org/doc/html/rfc2181#section-5.2) clarifies that the TTL of a record will be same for all IPs. Thus, whenever a partial set of IPs are received by different CoreDNS pods, the TTL values will be same.

I think the RFE is saying that the TTLs will be the same _at any given point in time_.  However, you will usually see TTLs decrement over time between queries.  Below is a quick example.  Are you accounting for that?  
<details>

Example:
```console
% repeat 3; do dig @8.8.8.8 www.google.com ; sleep 3; done

; <<>> DiG 9.11.36-RedHat-9.11.36-8.el8 <<>> @8.8.8.8 www.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 8371
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;www.google.com.                        IN      A

;; ANSWER SECTION:
www.google.com.         108     IN      A       64.233.176.105
www.google.com.         108     IN      A       64.233.176.103
www.google.com.         108     IN      A       64.233.176.147
www.google.com.         108     IN      A       64.233.176.104
www.google.com.         108     IN      A       64.233.176.106
www.google.com.         108     IN      A       64.233.176.99

;; Query time: 64 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Thu Aug 03 04:43:09 EDT 2023
;; MSG SIZE  rcvd: 139


; <<>> DiG 9.11.36-RedHat-9.11.36-8.el8 <<>> @8.8.8.8 www.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 43282
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;www.google.com.                        IN      A

;; ANSWER SECTION:
www.google.com.         105     IN      A       64.233.176.105
www.google.com.         105     IN      A       64.233.176.103
www.google.com.         105     IN      A       64.233.176.147
www.google.com.         105     IN      A       64.233.176.104
www.google.com.         105     IN      A       64.233.176.106
www.google.com.         105     IN      A       64.233.176.99

;; Query time: 34 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Thu Aug 03 04:43:12 EDT 2023
;; MSG SIZE  rcvd: 139


; <<>> DiG 9.11.36-RedHat-9.11.36-8.el8 <<>> @8.8.8.8 www.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 33288
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;www.google.com.                        IN      A

;; ANSWER SECTION:
www.google.com.         102     IN      A       64.233.176.105
www.google.com.         102     IN      A       64.233.176.103
www.google.com.         102     IN      A       64.233.176.147
www.google.com.         102     IN      A       64.233.176.104
www.google.com.         102     IN      A       64.233.176.106
www.google.com.         102     IN      A       64.233.176.99

;; Query time: 28 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Thu Aug 03 04:43:15 EDT 2023
;; MSG SIZE  rcvd: 139

% 
```
</details>

Reply:
The next lookup time (TTL + last lookup time) is used to check if the expiration time of the IP addresses match or not.